### DISCUSSION TITLE
Optionally Including 'select' and 'include'

### DISCUSSION THREAD

**Discussion by Bradleykingz** [prisma/prisma#2603] [Q&A] [answered]
I've gotten myself in a bind trying to create a method that works on top of Prisma's API. The problem I have is trying to create a single method to handle both 'select' and 'include',

From my understanding, you can't use both 'select' and 'include' together.  So, in order to create such a method, I used [this solution](https://stackoverflow.com/a/47892178/6701776).

Here's my method:

```js
async create(
    data: UserType,
    select?: UserSelect,
    include?: UserInclude
  ): Promise<UserType> {
    await super.create(data, select, include);
    this.log
      .context("createNewUser")
      .info("Called method with params", data, select, include);

      return prisma.user.create({
          data: {
              profile: {

                  connect: {id: data.profile.id},
              },
              account: {
                  connect: {id: data.account.id},
              },
          },
          ...(include && !select && {include}),
          ...(select && !include && {select}),
      });
  }
```

The idea is that select and include will never be used together.

However, on running this code with mocha, I get the following error:

```
Please either use `include` or `select`, but not both at the same time.
```

Is there a problem with my implementation or potentially a better way to go about this?

**Comment by ryands17** [comment]
Hey @Bradleykingz :wave: 

Here's a function that I have created regarding your use case:

```ts
async function main(select?: Prisma.UserSelect, include?: Prisma.UserInclude) {
  await prisma.user.create({
    data: {
      email: 'u1@g.com',
      password: 'user1',
    },
    ...(!select && include && { include }),
    ...(!include && select && { select }),
  })
}

main({
  name: true,
})
```

I have just included `select` and it works fine.

**Reply by Bradleykingz** [reply]
I'm not sure what method you used to run it, but with mocha + typescript-node, I'm still getting the same error.

This is what your function looks like in my IDE.

![Screenshot from 2020-05-29 16-42-05](https://user-images.githubusercontent.com/19578478/83266368-68fc0000-a1cb-11ea-9403-a007bf83a1ce.png)

edit: looks like it works as long as there's no return type (or the return type is 'any'), but that's not optimal.

**Reply by ryands17** [reply]
What version of Prisma are you using?
My current version of `@prisma/client` and `@prisma/cli` is `2.0.0-beta.6`. If so, could you try it with the latest version?

And still if the above doesn't work, could you share your setup so I could run mocha locally?

Also I notice that you are getting errors directly in the Editor. I am not getting that, so I guess a version update should do the trick.

**Reply by Bradleykingz** [reply]
I've upgraded from the alpha version to beta-6 but the error persists.

I'm starting to think it's more of a Typescript issue. I get the same error when I try to compile the ts code manually:

```
src/db/dao/User.ts:72:9 - error TS2322: Type '"Please either choose `select` or `include`"' is not assignable to type 'UserType'.

 72         return  prisma.user.create({
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 73             data: {
    ~~~~~~~~~~~~~~~~~~~
... 
 78             ...(!include && select && {select}),
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 79         })
    ~~~~~~~~~~

```

**Reply by ryands17** [reply]
@Bradleykingz yes you're correct that it's a TypeScript one. Typescript cannot infer the type if you pass in dynamic values as arguments, for the above to work, you would have to use something like function overloading in TS to map the types based on the arguments.

**Reply by Bradleykingz** [reply, marked as answer]
That pointed me in the right direction.

Thanks for the help!

**Comment by Bradleykingz** [comment]
For documentation purposes, this was one of the mistakes I was making:

'role' references account with a 1:1 relationship, so

```js
      return userDAO.create(
          {
            account,
            profile,
          },
          {
            account: {
              select: {
                email: true,
                isVerified: true,
                username: true,
                uuid: true,
              },
              include: {
                role: true,
              },
            },
          }
      );
```
However. This results in the following error:

```bash
Please either choose `select` or `include not both at the same time
```
I changed it to: 

```js
return userDAO.create(
          {
            account,
            profile,
          },
          {
            account: {
              select: {
                email: true,
                isVerified: true,
                username: true,
                uuid: true,
                role: true,
              },
            },
          }
      );
```
And now it works.

But IMO, the way the error message is formatted isn't very intuitive.

**Reply by ryands17** [reply]
@Bradleykingz I think what you were trying to achieve was either passing `select` or `include`. But in the above, you were passing both, which is what the message was trying to convey.

If you want to keep it optional as discussed above using TypeScript then that would be possible.

**Reply by pavelruzicka** [reply]
Hi Ryan, sorry to bring this year-old discussion back to life but I had some thoughts to share on this error message as I'd recently run into it and it had me very confused.

In my opinion, which isn't very different to Bradley's, the formatting of the message isn't all too intuitive and doesn't explain to the user that they can probably achieve their desired result with using only one of `include` and `select`. Obviously, this isn't what error messages are generally for and I would expect a write-up on this as part of the documentation (and if there is, obviously this whole comment can and should be scrapped).

Thank you for considering whether the wording could not be improved – personally, I was stuck with it until I managed to stumble upon this very discussion.

**Reply by ryands17** [reply]
@pavelruzicka 👋 

If you're using TypeScript, you already get an error in your code before execution that `select` and `include` cannot be used together.

![image](https://user-images.githubusercontent.com/19697099/120749079-8a4fa480-c521-11eb-8638-53a1b0fd2428.png)

And also the error message is clear in the sense that only one of them can be used at a time:

![image](https://user-images.githubusercontent.com/19697099/120749132-9cc9de00-c521-11eb-8a34-1fce25bd189b.png)

**Reply by itsmuaaz** [reply]
For some reason, vscode does not give me an error when I have both `select` and `include`, I only see the error in runtime from Prisma.

**Reply by yvipinaxioned** [reply]
Same here. No errors in IDE. I'm using vscode prisma extension too.

**Reply by brunobely** [reply]
Late to the party here but I just faced this error for the first time (did not get a warning in VS Code) and it's not clear at all to me why `select` also does `include`'s job -- if `select` picks scalar fields, and `include` picks relations, IMO the better API is to allow both to be used together rather than to `select` relations if trying to retrieve a partial result.
